File: go/CountSubMatriceCond/CountSubMatriceCond.go
package main

import "fmt"

// Time -> 0(max(N, M)) Space -> 0(1)
func numberOfSubmatrices(grid [][]byte) int {
	var count int
	countX := 0
	countY := 0

	bottom, right := 0, 0
	for bottom < len(grid) || right < len(grid[0]) {
		if right == len(grid[0]) {
			right = len(grid[0]) - 1
		}

		if bottom == len(grid) {
			bottom = len(grid) - 1
		}

		for i := 0; i <= bottom; i++ {
			count += check(i, right, &countX, &countY, grid)
		}
		bottom++
		right++
	}

	return count
}

func check(i, j int, countX, countY *int, grid [][]byte) int {
	val := grid[i][j]
	if val == 'X' {
		*countX++
	}

	if val == 'Y' {
		*countY++
	}

	if *countX == *countY && *countX != 0 {
		return 1
	}
	return 0
}

func main() {
	fmt.Println(numberOfSubmatrices([][]byte{{'X', 'Y', '.'}, {'Y', '.', '.'}}))
	fmt.Println(numberOfSubmatrices([][]byte{{'X', 'X'}, {'X', 'Y'}}))
	fmt.Println(numberOfSubmatrices([][]byte{{'.', '.'}, {'Y', 'X'}}))
}

File: python/countSubMatriceCond.py
# Time -> 0(max(N, M)) Space -> 0(1)
def numberOfSubmatrices(grid: list[list[str]]) -> int:
    count = 0
    x_count = 0
    y_count = 0

    for val in grid[0]:
        if val == "X":
            x_count += 1
        elif val == "Y":
            y_count += 1

        if x_count == y_count and x_count != 0:
            count += 1

    x_count = 0
    y_count = 0
    for idx in range(len(grid)):
        val = grid[0][idx]
        if val == "X":
            x_count += 1
        elif val == "Y":
            y_count += 1

        if x_count == y_count and x_count != 0:
            count += 1

    return count


print(numberOfSubmatrices([["X", "Y", "."], ["Y", ".", "."]]))
print(numberOfSubmatrices([["X", "X"], ["X", "Y"]]))
print(numberOfSubmatrices([[".", "."], [".", "."]]))
